java-checks/src/main/java/org/sonar/java/checks/SimpleClassNameCheck.java

```diff
@@ -0,0 +1,121 @@
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import org.sonar.check.Rule;
+import org.sonar.java.model.SyntacticEquivalence;
+import org.sonar.java.resolve.SemanticModel;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.JavaFileScannerContext;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.ImportTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.Tree.Kind;
+
+@Rule(key = "S1942")
+public class SimpleClassNameCheck extends IssuableSubscriptionVisitor {
+
+  private static final String MESSAGE = "Replace this fully qualified name with \"%s\"";
+  private Set<Tree> hasIssue = new HashSet<>();
+  private Set<Tree> importedWithWildcardNames = new HashSet<>();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Kind.IMPORT, Kind.MEMBER_SELECT);
+  }
+
+  @Override
+  public void scanFile(JavaFileScannerContext context) {
+    super.scanFile(context);
+    hasIssue.clear();
+    importedWithWildcardNames.clear();
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+
+    if (tree.is(Kind.IMPORT)) {
+      visitImport((ImportTree)tree);
+
+    } else {
+      visitMemberSelect((MemberSelectExpressionTree)tree);
+    }
+  }
+
+  private void visitMemberSelect(MemberSelectExpressionTree tree) {
+    if (!isInImport(tree) && !hasIssue.contains(tree) && !tree.identifier().symbol().isPackageSymbol()) {
+
+      for (Tree importedWildcard : importedWithWildcardNames) {
+        if (SyntacticEquivalence.areEquivalent(importedWildcard, tree.expression())) {
+          reportIssue(tree, String.format(MESSAGE, tree.identifier().name()));
+          return;
+        }
+
+      }
+    }
+
+  }
+
+  private static boolean isInImport(Tree tree) {
+    Tree parent = tree.parent();
+
+    while (parent != null) {
+      if (parent.is(Kind.IMPORT)) {
+        return true;
+      }
+      parent = parent.parent();
+    }
+
+    return false;
+  }
+
+  private void visitImport(ImportTree tree) {
+    SemanticModel semanticModel = (SemanticModel) context.getSemanticModel();
+    Symbol symbol = semanticModel.getSymbol(tree);
+
+    if (symbol != null) {
+      checkImportedSymbol(symbol);
+
+    } else {
+      importedWithWildcardNames.add(((MemberSelectExpressionTree)tree.qualifiedIdentifier()).expression());
+    }
+  }
+
+  private void checkImportedSymbol(Symbol symbol) {
+    for (IdentifierTree usageIdentifier : symbol.usages()) {
+      Tree parent = usageIdentifier.parent();
+
+      if (parent.is(Kind.MEMBER_SELECT) && ((MemberSelectExpressionTree) parent).expression().is(Kind.MEMBER_SELECT)) {
```

Comment:
This approach won't work if all the classes of a project are in a single root package:

```
package org;

import org.B;

class A {
  org.B b; // should be noncompliant
}
```

```
package org;

class B {
  ...
}
```

You probably should rather rely on the type of the symbol associated to the identifier of the parent member select, if the parent is a member select.

```
if (parent.is(Kind.MEMBER_SELECT)) {
  ExpressionTree expression = ((MemberSelectExpressionTree) parent).expression();
  if (expression.is(Kind.MEMBER_SELECT) && !((MemberSelectExpressionTree) expression).identifier().symbol().isTypeSymbol()) {
    return;
  }
  reportIssue(parent, String.format(MESSAGE, symbol.name()));
  hasIssue.add(parent);
}
```